src/Swaps/Swapper.sol

```diff
@@ -0,0 +1,134 @@
+    error Swapper_UnknownHandler();
+    error Swapper_Reentrancy();
+    error Swapper_InsufficientBalance();
+    error Swapper_PastDeadline();
+
+    // In the locked state, allow contract to call itself, but block all external calls
+    modifier externalLock() {
+        bool isExternal = msg.sender != address(this);
+
+        if (isExternal) {
+            if (reentrancyLock == REENTRANCYLOCK_LOCKED) revert Swapper_Reentrancy();
+            reentrancyLock = REENTRANCYLOCK_LOCKED;
+        }
+
+        _;
+
+        if (isExternal) reentrancyLock = REENTRANCYLOCK_UNLOCKED;
+    }
+
+    constructor(address oneInchAggregator, address uniswapRouterV2, address uniswapRouterV3, address uniswapRouter02)
+        OneInchHandler(oneInchAggregator)
+        UniswapV2Handler(uniswapRouterV2)
+        UniswapV3Handler(uniswapRouterV3)
+        UniswapAutoRouterHandler(uniswapRouter02)
+    {}
+
+    /// @inheritdoc ISwapper
+    function swap(SwapParams memory params)
+        public
+        externalLock
+    {
+        if (params.mode >= MODE_MAX_VALUE) revert Swapper_UnknownMode();
+        if (params.deadline < block.timestamp) revert Swapper_PastDeadline();
+
+        if (params.handler == HANDLER_ONE_INCH) {
+            swapOneInch(params);
+        } else if (params.handler == HANDLER_UNISWAP_V2) {
+            swapUniswapV2(params);
+        } else if (params.handler == HANDLER_UNISWAP_V3) {
+            swapUniswapV3(params);
+        } else if (params.handler == HANDLER_UNISWAP_AUTOROUTER) {
+            swapAutoRouter(params);
+        } else {
+            revert Swapper_UnknownHandler();
+        }
+
+        if (params.mode == MODE_EXACT_IN) return;
+
+        // swapping to target debt is only useful for repaying
+        if (params.mode == MODE_TARGET_DEBT) {
+            // at this point amountOut holds the required repay amount
+            repayAndDeposit(params.tokenOut, params.receiver, params.amountOut, params.account);
+        }
+
+        // return unused input token after exact output swap
+        deposit(params.tokenIn, params.vaultIn, 0, params.account);
+    }
+
+    /// @inheritdoc ISwapper
+    /// @dev in case of over-swapping to repay, pass max uint amount
+    function repay(address token, address vault, uint256 repayAmount, address account) public externalLock {
+        uint256 balance = setMaxAllowance(token, vault);
+        if (repayAmount != type(uint256).max && repayAmount > balance) revert Swapper_InsufficientBalance();
+
+        IEVault(vault).repay(repayAmount, account);
+    }
+
+    /// @inheritdoc ISwapper
+    function repayAndDeposit(address token, address vault, uint256 repayAmount, address account) public externalLock {
+        uint256 balance = setMaxAllowance(token, vault);
+        if (repayAmount != type(uint256).max && repayAmount > balance) revert Swapper_InsufficientBalance();
+
+        IEVault(vault).repay(repayAmount, account);
+
+        if (balance > repayAmount) {
```

Comment:
this won't work if `repayAmount == type(uint256).max` and some dust is kept after the repayment

Reply:
this comment was hidden, so maybe no need to reply.. good find, but I'd keep as is - cheaper for the common use case of the funtion for swap and repay mode, where `repayAmount` won't be max. If max is needed, we can use `repay` and then `deposit`